packages/vscode-extension/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { configure, getLogger, Logger } from "log4js";
+import * as os from "os";
+import * as path from "path";
+// eslint-disable-next-line import/no-unresolved
+import * as vscode from "vscode";
+
+import {
+  ConfigFolderName,
+  FxError,
+  Stage,
+  TelemetryReporter,
+  UserError,
+} from "@microsoft/teamsfx-api";
+import { Correlator } from "@microsoft/teamsfx-core";
+import Reporter from "@vscode/extension-telemetry";
+
+import * as extensionPackage from "../../package.json";
+import * as globalVariables from "../globalVariables";
+import {
+  FeatureFlags,
+  getAllFeatureFlags,
+  getProjectId,
+  isFeatureFlagEnabled,
+} from "../utils/commonUtils";
+import { TelemetryCache } from "./cache";
+import {
+  TelemetryComponentType,
+  TelemetryErrorType,
+  TelemetryEvent,
+  TelemetryEventCache,
+  TelemetryProperty,
+  TelemetrySuccess,
+} from "./extTelemetryEvents";
+
+const TelemetryTestLoggerFile = "telemetryTest.log";
+
+/* eslint-disable prefer-const */
+let isFromSample: boolean | undefined = undefined;
+let settingsVersion: string | undefined = undefined;
+let isM365: boolean | undefined = undefined;
+
+export let reporter: VSCodeTelemetryReporter;
+// export for UT
+export let lastCorrelationId: string | undefined = undefined;
+
+/**
+ *  VSCode telemetry reporter used by fx-core and extension.
+ *  Usage:
+ *    let reporter = new VSCodeTelemetryReporter(key, extensionVersion, extensionId)
+ *  Illustrate:
+ *    key = <'the application insights key'>, 'aiKey' in package.json
+ *    extensionVersion = '<extension version>', extension version will be reported as a property with each event
+ *    extensionId = '<your extension unique name>', all events will be prefixed with this event name. eg: 'extensionId/eventname'
+ */
+export class VSCodeTelemetryReporter extends vscode.Disposable implements TelemetryReporter {
```

Comment:
Since VSCodeTelemetryReport itself is also vscode.Disposable, should we dispose it too?

Reply:
It is disposed in `dispose()` function which is invoked in `deactivate()`

Reply:
VSCodeTelemetryReporter.dispose() is called in extension.deactivate() which only disposes this.reporter, right?

Reply:
VSCodeTelemetryReporter.dispose() does 3 things:
1. send events in cache.
2. persist events on disk.
3. dispose `this.reporter`.